[CPU] enable CI for PRs, add Dockerfile and auto build task

ZailiWang opened the pull request:
<!-- Thank you for your contribution! We appreciate it. The following guidelines will help improve your pull request and facilitate feedback. If anything is unclear, don't hesitate to submit your pull request and ask the maintainers for assistance. -->

## Motivation

The PR is for enabling the docker env setup and CI processes for running SGLang on Xeon CPU servers.

## Modifications

Add the dockerfile, and yml files for CPU part of CI process and auto docker image build-up.
The test files are updated to enable device specific tests, as well as the CPU test cases.

## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

DiweiSun committed:
Create Dockerfil.xeon

DiweiSun committed:
Rename Dockerfil.xeon to Dockerfile.xeon

DiweiSun committed:
enable device auto detection for cuda/cpu/intel-gpu/rocm

DiweiSun committed:
Delete docker/Dockerfile.xeon

DiweiSun committed:
lint format fix

DiweiSun committed:
Merge branch 'main' into molly/cpu_ut

ZailiWang committed:
add Dockerfile for Xeon

ZailiWang commented on docker/Dockerfile.xeon:
```diff
@@ -0,0 +1,43 @@
+    pip install -e "python[all_cpu]" && \
+    pip install torch==${VER_TORCH} torchvision==${VER_TORCHVISION} --index-url https://download.pytorch.org/whl/cpu --force-reinstall && \
+    cd sgl-kernel && \
+    python setup_cpu.py install
```
Will update to use CMake when the resolving PR is merged.

mingfeima replied:
could you ping the PR link you mentioned here and mark this one as `draft` for now

ZailiWang replied:
Sure. It's #6419

mingfeima replied:
> Sure. It's #6419

landed.

ZailiWang replied:
updated to use CMake.

DiweiSun committed:
Update run_suite.py

mingfeima commented:
can docker file accepts arguements? if yes, we can change TORCH_VERSION as an arguement and set it default to be 2.6

ZailiWang commented:
> can docker file accepts arguements? if yes, we can change TORCH_VERSION as an arguement and set it default to be 2.6

Yeah, by setting `ARG TORCH_VERSION=2.6.0` in the Dockerfile, it is a settable argument and can be changed by e.g. `--build-arg TORCH_VERSION=2.7.0`  in `docker build` command

ZailiWang committed:
Merge branch 'sgl-project:main' into main

ZailiWang committed:
add autotask yml file

ZailiWang committed:
Merge branch 'main' into main

ZailiWang committed:
Merge branch 'main' into main

ZailiWang committed:
Merge branch 'main' into main

ZailiWang committed:
Merge branch 'main' into main

ZailiWang committed:
replace setup.py

ZailiWang committed:
Merge branch 'main' into main

ZailiWang committed:
fix dep. issue in test_topk

DiweiSun committed:
Merge branch 'main' into molly/cpu_ut

DiweiSun committed:
Merge pull request #5 from ZailiWang/main

xeon docker build

ZailiWang committed:
install torch_cpu separately since --index-url cannot be set in toml

ZailiWang committed:
Merge branch 'main' into main

DiweiSun committed:
enable cpu ci

DiweiSun committed:
docker build with local hardware

DiweiSun committed:
docker build with local hardware

DiweiSun committed:
bugfix for workspace

DiweiSun committed:
bugfix for container clean

DiweiSun committed:
remove test_rope.py in cpu ci

DiweiSun committed:
remove proxy setting and lint format fix

DiweiSun committed:
Merge branch 'main' into molly/cpu_ut

DiweiSun committed:
UT TEST: raise error when server launch process dead

DiweiSun committed:
enable ci on xeon

test on self-hosted action runners, all passed. https://github.com/DiweiSun/sglang/actions/runs/15269185632

ZailiWang committed:
skip CPU test if AMX not supported

ZailiWang committed:
Merge branch 'main' into main

mingfeima commented:
@ZailiWang  change the tile and provide some details of what this PR is for?

ZailiWang committed:
fix lint error

ZailiWang committed:
Merge branch 'main' of https://github.com/ZailiWang/sglang

ZailiWang commented:
> @ZailiWang can we satisfy:
> 
> * skip sgl-kernel cpu tests when CPU doesn't have AMX
> * skip certain tests when vllm is not installed (since some test cases, e.g. rope, uses reference from sglang)

Hi @zhyncs ,
We have drafted this PR for CPU CI and docker env build-up. Yet there are still something that we are unsure. Your comment and guidance is highly appreciated!

- For satisfying this
> * skip sgl-kernel cpu tests when CPU doesn't have AMX

We added a `Check AMX Support` step in CPU CI process and CPU test cases are triggered only when the check is passed [here](https://github.com/sgl-project/sglang/pull/6458/files#diff-8db469c7176c54d966f2aaa46060f34d2f0ecb2a1463e6bd36ec11e9bd60c82eR51). Is this good or would you comment with some better solution?

- For CPU docker build-up, we created `release-docker-xeon.yml` [here](https://github.com/sgl-project/sglang/pull/6458/files#diff-05c755c8d2d45154d9a59cd1324f93dc96dc083d5900e026d37aaaf1e2876829) by imitating "release-docker" files for other devices. Please correct if anything still required to be updated in this file.

- In 2 CPU test cases (test_rope.py and test_topk.py), `vllm` implementation is still invoked when calculating the reference results. Since in SGLang we do not install vllm as a dependency now, such errors would be thrown:

```
File "/sgl-workspace/sglang/python/sglang/srt/layers/quantization/utils.py", line 14, in <module>
    from vllm._custom_ops import scaled_fp8_quant
ModuleNotFoundError: No module named 'vllm'
```

```
File "/sgl-workspace/sglang/python/sglang/srt/layers/rotary_embedding.py", line 87, in __init__
    from vllm._custom_ops import rotary_embedding
ModuleNotFoundError: No module named 'vllm'
```

Hence these 2 test cases are not added into `test_suite.py` yet.

zhyncs committed:
Merge branch 'main' into main

zhyncs committed:
Merge branch 'main' into main

zhyncs committed:
Merge branch 'main' into main

ZailiWang commented:
> In 2 CPU test cases (test_rope.py and test_topk.py), vllm implementation is still invoked when calculating the reference results. Since in SGLang we do not install vllm as a dependency now, such errors would be thrown:

We have other PRs to mitigate vllm dependency issue now.

zhyncs commented on .github/workflows/pr-test-xeon.yml:
```diff
@@ -0,0 +1,86 @@
+jobs:
+  build-test:
+    if: github.event_name == 'pull_request'
+    runs-on: spr-node
```
Currently this has not been set up in the sglang repo as a self-hosted runner.

zhyncs commented on .github/workflows/pr-test-xeon.yml:
```diff
@@ -0,0 +1,86 @@
+  finish:
+    if: always()
+    needs: [build-test, unit-test]
+    runs-on: spr-node
```
```suggestion
    runs-on: ubuntu-24.04
```

zhyncs commented on .github/workflows/pr-test-xeon.yml:
```diff
@@ -0,0 +1,86 @@
+  unit-test:
+    if: github.event_name == 'pull_request'
+    needs: [build-test]
+    runs-on: spr-node
```
```suggestion
    runs-on: sgl-kernel-build-node
```

zhyncs committed:
Merge branch 'main' into main

zhyncs committed:
upd

zhyncs committed:
upd

zhyncs committed:
upd
